Initial draft of model explanations chapter (#151)
tidymodels/TMwR: 18-explaining-models-and-predictions.Rmd

- `r pkg(DALEX)` functions when we want to use _model-agnostic_ methods that can be applied to any model. 

In Chapters \@ref(resampling) and \@ref(compare), we trained and compared several models to predict the price of homes in Ames, IA, including a linear model with interactions and a random forest model.

```{r include=FALSE}
bind_rows(
  augment(lm_fit, ames_train) %>% mutate(model = "lm + interactions"),
  augment(rf_fit, ames_train) %>% mutate(model = "random forest")
) %>%
  ggplot(aes(Sale_Price, .pred, color = model)) +
  geom_abline(col = "gray50", lty = 2) + 
  geom_point(alpha = 0.5, show.legend = FALSE) +
  facet_wrap(vars(model))
```

Let's build model-agnostic explainers for both of these models to why they make the predictions they do. We can use the `r pkg(DALEXtra)` add-on package for `r pkg(DALEX)`, which provides support for tidymodels. @Biecek2021 provide a thorough exploration of how to use `r pkg(DALEX)` for model explanations; this chapter only summarizes some important approaches, specific to tidymodels. To compute any kind of model explanation, global or local, using `r pkg(DALEX)`, we first create an _explainer_ for each model.

```{r message=FALSE}

> library(DALEXtra)

Reviewer: Should we also add

```
conflicted::conflict_prefer("vi", "vip")
conflicted::conflict_prefer("explain", "lime")
```

> library(DALEXtra)

Author: I actually ended up using all DALEX so far in the chapter so I don't think it's necessary at this point. I could remove them from the `include = FALSE` chunk. These are examples of functions that don't work so well if you call `tidymodels_prefer()`, though; it's one of those situations where things are happily going along before you call `tidymodels_prefer()` and then functions like these don't work anywhere afterwards.

Maybe if in the beans section we use vip we can use it, [link back to here](https://www.tmwr.org/base-r.html#the-tidymodels-metapackage), and use it as an example of the pros/cons of opting in to conflicted.

vip_features <- c("Neighborhood", "Gr_Liv_Area", "Year_Built", 
                  "Bldg_Type", "Latitude", "Longitude")

vip_train <- ames_train %>% 
  select(all_of(vip_features))

explainer_lm <- explain_tidymodels(
  lm_fit, 
  data = vip_train, 
  y = ames_train$Sale_Price,
  label = "lm + interactions",
  verbose = FALSE
)

explainer_rf <- explain_tidymodels(
  rf_fit, 
  data = vip_train, 
  y = ames_train$Sale_Price,
  label = "random forest",
  verbose = FALSE
)
```

:::rmdwarning
A linear model is typically straightforward to interpret and explain; you may not often find yourself computing separate model explanations for a linear model. However, it can sometimes be difficult to understand or explain the predictions of even a linear model once it has splines and interaction terms!
:::

> 

Reviewer: Can you talk up-front about what is being measures for importance? 

Maybe use terminology like: 

- _basic predictors_: the original data columns (perhaps with simple transformations such as the log, etc). 
- _derived features_: columns created in the data set form the original predictors that are calculated from one or more of the basic predictors (e.g. spline terms, PCA components, ratios)

I don't recall seeing existing terminology in the DALEX book (but maybe I missed it).

## Local explanations

Local model explanations provide information about a prediction for a single observation. For example, let's consider an older duplex in the North Ames neighborhood (Section \@ref(model-types)).

```{r}
duplex <- vip_train[121,]
duplex
```

There are multiple possible approaches to understanding why a model predicts a given price for this duplex. One is called a "break-down" explanation, and computes how contributions attributed to individual features change the mean model's prediction for a particular observation, like our duplex. For the linear model, the duplex status, size, longitude, and age all contribute the most to the price being driven down from the intercept.

```{r}
lm_breakdown <- predict_parts(explainer = explainer_lm, new_observation = duplex)
lm_breakdown
```

> The most important features are slightly different for the random forest model, with the size, age, and duplex status being most important.

Reviewer: We should also comment here that the effect of `Latitude` combines the effects of all of its individual spline terms that were in the model.

  geom_col(data = ~distinct(., variable, mean_val), 
           aes(mean_val, variable), 
           alpha = 0.5) +
  geom_boxplot(width = 0.5) +
  theme(legend.position = "none") +
  labs(y = NULL)
```

What about a different observation in our data set? Let's look at a larger, newer one-family home in the Gilbert neighborhood.

```{r}
big_house <- vip_train[1177,]
big_house
```

We compute SHAP average attributions in the same way.

```{r}
shap_house <- 
  predict_parts(explainer = explainer_rf, 
                new_observation = big_house, 
                type = "shap",
                B = 20)
```

```{r echo=FALSE}
shap_house %>%
  group_by(variable) %>%
  mutate(mean_val = mean(contribution)) %>%
  ungroup() %>%
  mutate(variable = fct_reorder(variable, abs(mean_val))) %>%
  ggplot(aes(contribution, variable, fill = mean_val > 0)) +
  geom_col(data = ~distinct(., variable, mean_val), 
           aes(mean_val, variable), 
           alpha = 0.5) +
  geom_boxplot(width = 0.5) +
  theme(legend.position = "none") +
  labs(y = NULL)
```

Unlike the duplex, the size and age of this house contribute to its price being higher.

## Global explanations

> Global model explanations, also called global feature importance or variable importance, help us understand which features are most important in driving the predictions of these two models overall, aggregated over the whole training set. While the previous section addressed what variables or features are most important in predicting sale price for an individual home, global feature importance addresses what variables are most important for a model overall.

Reviewer: ```suggestion
Global model explanations, also called global feature importance or variable importance, help us understand which features are most important in driving the predictions of these two models overall, aggregated over the whole training set. While the previous section addressed what variables or features are most important in predicting sale price for an individual home, global feature importance addresses what variables are most important for a in aggregate.
```

:::rmdnote
One way to compute variable importance is to _permute_ the features [@breiman2001random]. We can permute or shuffle the values of a feature, predict from the model, and then measure how much worse the model fits the data compared to before shuffling.
:::

> If shuffling a feature causes a large degradation in model performance, it is an important feature; if shuffling a feature's values doesn't make much difference to how the model performs, it must not be an important feature. This approach can be applied to any kind of model (it is _model-agnostic_) and the results are straightforward to understand.

Reviewer: ```suggestion
If shuffling a column causes a large degradation in model performance, it is important; if shuffling a column's values doesn't make much difference to how the model performs, it must not be an important variable. This approach can be applied to any kind of model (it is _model-agnostic_) and the results are straightforward to understand.
```

Just an example of reserving the word feature for a derived value.

Using `r pkg(DALEX)`, we compute this kind of variable importance via the `model_parts()` function.

```{r}
vip_lm <- model_parts(explainer_lm, loss_function = loss_root_mean_square) 
vip_rf <- model_parts(explainer_rf, loss_function = loss_root_mean_square)
```

Again, we could use the default plot method from `r pkg(DALEX)` by calling `plot(vip_lm, vip_rf)` but the underlying data is available for exploration, analysis, and plotting.

```{r}
full_vip <- bind_rows(vip_lm, vip_rf) %>%
  filter(variable != "_baseline_")

full_rmse <- full_vip %>% 
  filter(variable == "_full_model_") %>% 
  group_by(label) %>% 
  summarise(dropout_loss = mean(dropout_loss))

full_vip %>%
  filter(variable != "_full_model_") %>% 

>   mutate(variable = forcats::fct_reorder(variable, dropout_loss)) %>%

Reviewer: ```suggestion
  mutate(variable = fct_reorder(variable, dropout_loss)) %>%
```

  ggplot(aes(dropout_loss, variable, color = label, fill = label)) +
  geom_vline(data = full_rmse, aes(xintercept = dropout_loss, color = label),
             size = 1.4, lty = 2, alpha = 0.7) +
  geom_boxplot(alpha = 0.2) +
  facet_wrap(~label, ncol = 1) +
  theme(legend.position = "none") +

>   labs(x = "RMSE after permutations (higher indicates more important)", y = NULL, fill = NULL, color = NULL)

Reviewer: ```suggestion
  labs(x = "RMSE after permutations (higher indicates more important)",
       y = NULL,
       fill = NULL,
       color = NULL)
```

## Building global explanations from local explanations

So far in this chapter, we have focused on local model explanations for a single observation (via Shapley additive explanations) and global model explanations for a data set as a whole (via permuting features). It is also possible to build global model explanations up by aggregating local model explanations, as with _partial dependence profiles_.

:::rmdnote
Partial dependence profiles show how the expected value of a model prediction, like the predicted price of a home in Ames, changes as a function of a feature, like the age or gross living area.
:::

One way to build such a profile is by aggregating or averaging profiles for individual observations. A profile showing how an individual observation's prediction changes as a function of a given feature is called an ICE (individual conditional expectation) profile or a CP (*ceteris paribus*) profile. We can compute such individual profiles (for 500 of the observations in our training set) and then aggregate them using the `r pkg(DALEX)` function `model_profile()`.

```{r}
pdp_age <- model_profile(explainer_rf, N = 500, variables = "Year_Built")
```

> 

Reviewer: This kinds of makes it look like `model_profile()~ produces the plot. Maybe show the plot code?

Reviewer: Since we use very similar code a few times, I'd make a function for this.

```{r echo=FALSE}
as_tibble(pdp_age$agr_profiles) %>%
  ggplot(aes(`_x_`, `_yhat_`)) +
  geom_line(data = as_tibble(pdp_age$cp_profiles),
            aes(x = Year_Built, group = `_ids_`),
            size = 0.5, alpha = 0.1, color = "gray50") +
  geom_line(size = 1.2, alpha = 0.8, color = "midnightblue") +
  labs(x = "Year built", 
       y = "Sale Price (log)", 
       color = NULL)
```

> Partial dependence profiles can be computed for any other feature in the model, and also for groups in the data, such as `Bldg_Type`. Let's use `1e3` observations for these profiles.

Reviewer: ```suggestion
Partial dependence profiles can be computed for any other feature in the model, and also for groups in the data, such as `Bldg_Type`. Let's use 1,000 observations for these profiles.
```

```{r}

> pdp_liv <- model_profile(explainer_rf, N = 1e3, 

Reviewer: ```suggestion
pdp_liv <- model_profile(explainer_rf, N = 1000, 
```

       color = NULL)
```

We have the option of using `plot(pdp_liv)`, but since we are making plots with the underlying data here, we can even facet by one of the features to visualize how the predictions change differently, highlighting the model's behavior for subgroups.

```{r echo=FALSE}
as_tibble(pdp_liv$agr_profiles) %>%
  mutate(Bldg_Type = stringr::str_remove(`_label_`, "random forest_")) %>%
  ggplot(aes(`_x_`, `_yhat_`, color = Bldg_Type)) +
  geom_line(data = as_tibble(pdp_liv$cp_profiles),
            aes(x = Gr_Liv_Area, group = `_ids_`),
            size = 0.5, alpha = 0.1, color = "gray50") +
  geom_line(size = 1.2, alpha = 0.8, show.legend = FALSE) +
  scale_x_log10() +
  facet_wrap(~Bldg_Type) +
  scale_color_brewer(palette = "Dark2") +
  labs(x = "Gross living area", 
       y = "Sale Price (log)", 
       color = NULL)
```

There is not one right approach for building model explanations and the options outlined in this chapter are not exhaustive. We point you to @Biecek2021 and @Molnar2021 for further reading, but highlight in this chapter good options for explanations at both the individual and global level, and how to bridge from one to the other.

> ## Beans beans beans

Reviewer: I'll open the PR and add this section shortly
